From a3436ab24a3f828b01e43722f9811b1e358c474f Mon Sep 17 00:00:00 2001
From: "Benjamin A. Beasley" <code@musicinmybrain.net>
Date: Mon, 28 Jul 2025 09:53:56 -0400
Subject: [PATCH 1/2] Adapt to API changes in dr_wav 0.14

Adds `Filesystem_Stream::InputStream::GetPosition()` to support the new
`onTell` callback for `drwav_init_ex`.
---
 src/decoder_drwav.cpp     | 10 +++++++++-
 src/filesystem_stream.cpp |  6 +++++-
 src/filesystem_stream.h   |  1 +
 3 files changed, 15 insertions(+), 2 deletions(-)

diff --git a/src/decoder_drwav.cpp b/src/decoder_drwav.cpp
index c48ca260d..4fa9a4f52 100644
--- a/src/decoder_drwav.cpp
+++ b/src/decoder_drwav.cpp
@@ -52,9 +52,17 @@ drwav_bool32 seek_func(void* userdata, int offset, drwav_seek_origin origin) {
 	return DRWAV_TRUE;
 }
 
+drwav_bool32 tell_func(void* userdata, drwav_int64* cursor) {
+	auto* f = reinterpret_cast<Filesystem_Stream::InputStream*>(userdata);
+
+	*cursor = f->GetPosition();
+
+	return DRWAV_TRUE;
+}
+
 bool DrWavDecoder::Open(Filesystem_Stream::InputStream stream_) {
 	this->stream = std::move(stream_);
-	init = drwav_init_ex(&handle, read_func, seek_func, nullptr, &this->stream, nullptr, DRWAV_SEQUENTIAL, nullptr) == DRWAV_TRUE;
+	init = drwav_init_ex(&handle, read_func, seek_func, tell_func, nullptr, &this->stream, nullptr, DRWAV_SEQUENTIAL, nullptr) == DRWAV_TRUE;
 	return init;
 }
 
diff --git a/src/filesystem_stream.cpp b/src/filesystem_stream.cpp
index 3aaf5a32c..349f8c9ec 100644
--- a/src/filesystem_stream.cpp
+++ b/src/filesystem_stream.cpp
@@ -52,13 +52,17 @@ std::string_view Filesystem_Stream::InputStream::GetName() const {
 std::streampos Filesystem_Stream::InputStream::GetSize() const {
 	if (!size_cached) {
 		size_cached = true;
-		auto cur_pos = rdbuf()->pubseekoff(0, std::ios_base::cur, std::ios_base::in);
+		auto cur_pos = GetPosition();
 		size = rdbuf()->pubseekoff(0, std::ios_base::end, std::ios_base::in);
 		rdbuf()->pubseekoff(cur_pos, std::ios_base::beg, std::ios_base::in);
 	}
 	return size;
 }
 
+std::streampos Filesystem_Stream::InputStream::GetPosition() const {
+	return rdbuf()->pubseekoff(0, std::ios_base::cur, std::ios_base::in);
+}
+
 void Filesystem_Stream::InputStream::Close() {
 	delete rdbuf();
 	set_rdbuf(nullptr);
diff --git a/src/filesystem_stream.h b/src/filesystem_stream.h
index fe1c2465d..090fd7ab1 100644
--- a/src/filesystem_stream.h
+++ b/src/filesystem_stream.h
@@ -39,6 +39,7 @@ namespace Filesystem_Stream {
 
 		std::string_view GetName() const;
 		std::streampos GetSize() const;
+		std::streampos GetPosition() const;
 		void Close();
 
 		template <typename T>
-- 
2.50.1


From 7fcefce7128c091f6c9483ba05e7af8f2c143573 Mon Sep 17 00:00:00 2001
From: Ghabry <gabriel+github@mastergk.de>
Date: Tue, 29 Jul 2025 10:08:19 +0200
Subject: [PATCH 2/2] dr_wav: Conditional support for old dr_wav 0.13.x

---
 src/decoder_drwav.cpp | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/src/decoder_drwav.cpp b/src/decoder_drwav.cpp
index 4fa9a4f52..376766556 100644
--- a/src/decoder_drwav.cpp
+++ b/src/decoder_drwav.cpp
@@ -62,7 +62,11 @@ drwav_bool32 tell_func(void* userdata, drwav_int64* cursor) {
 
 bool DrWavDecoder::Open(Filesystem_Stream::InputStream stream_) {
 	this->stream = std::move(stream_);
+#if DRWAV_VERSION_MINOR < 14
+	init = drwav_init_ex(&handle, read_func, seek_func, nullptr, &this->stream, nullptr, DRWAV_SEQUENTIAL, nullptr) == DRWAV_TRUE;
+#else
 	init = drwav_init_ex(&handle, read_func, seek_func, tell_func, nullptr, &this->stream, nullptr, DRWAV_SEQUENTIAL, nullptr) == DRWAV_TRUE;
+#endif
 	return init;
 }
 
-- 
2.50.1

